docs/docs/uploads.md

```diff
@@ -0,0 +1,733 @@
+
+</details>
+
+What this configures is:
+
+**A) CRUD operations**
+No need to do anything here, but you have to use processors to supply Prisma with data in the correct format.
+
+What this will ensure is:
+
+- when the record is deleted, the associated upload is removed from storage
+- when a record is updated, the associated upload file is also replaced
+
+...and negative cases such as:
+
+- saved uploads are removed if creation fails
+- saved uploads are removed if update fails (while keeping the original)
+
+**B) Result extensions**
+
+```ts title="api/src/services/profiles/profiles.ts"
+export const profile = async ({ id }) => {
+  // 👇 await the result from your prisma query
+  const profile = await db.profile.findUnique({
+    where: { id },
+  })
+
+  // Convert the avatar field (which was persisted as a path) to data uri string
+  // highlight-next-line
+  return profile?.withDataUri()
+}
+```
+
+:::tip
+It's very important to note limitations around what Prisma extensions can do:
+
+**a) The CRUD operations will not run on nested read and write operations** <br/>
+For example:
+
+```js
+const savedFiles = saveFile.inList(input.files)
+
+db.folder.update({
+  data: {
+    ...input,
+    files: {
+      // highlight-start
+      createMany: {
+        data: savedFiles, // if the createMany fails, the saved files will _not_ be deleted
+      },
+      // highlight-end
+    },
+  },
+  where: { id },
+})
+```
+
+**b) Result extensions are not available on relations.**
+
+You can often rewrite the query in a different way though. For example, when looking up files :
+
+```ts
+const filesViaRelation = await db.folder
+  .findUnique({ where: { id: root?.id } })
+  .files()
+
+const filesWhereQuery = await db.file.findMany({
+  where: {
+    folderId: root?.id,
+  },
+})
+
+// 🛑 Will not work, because files accessed via relation
+// highlight-next-line
+return filesViaRelation.map((file) => file.withSignedUrl())
+
+// ✅ OK, because direct lookup
+// highlight-next-line
+return filesWhereQuery.map((file) => file.withSignedUrl())
+```
+
+:::
+
+## Upload savers
+
+You'll also need a way to actually save the incoming `File` object to a file persisted on storage. In your services, you can use the pre-configured "savers" to convert Files to strings for Prisma to save into the database. The savers, and storage adapters configured in `api/src/lib/uploads` determine where the file is saved.
+
+```ts title="api/src/services/profiles/profiles.ts"
+// highlight-next-line
+import { saveFiles } from 'src/lib/uploads'
+
+export const updateProfile: MutationResolvers['updateProfile'] = async ({
+  id,
+  input,
+}) => {
+  // highlight-next-line
+  const processedInput = await saveFiles.forProfile(input)
+
+  // input.avatar (File) becomes a string 👇
+  // The configuration on where it was saved is passed when we setup uploads in src/lib/uploads.ts
```

Comment:
I had to do a double-take on this sentence. There's nothing wrong with it, so feel free to keep it as-is, but here's a suggestion that I personally think makes it easier to read:

```suggestion
  // Configuration in src/lib/uploads.ts determine where the upload is saved
```

Or maybe

```suggestion
  // Settings in src/lib/uploads.ts configures where the upload is saved
```